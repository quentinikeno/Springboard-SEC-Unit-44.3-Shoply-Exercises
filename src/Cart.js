import { useSelector, shallowEqual } from "react-redux";
import CartItem from "./CartItem";

const Cart = () => {
	const { products } = useSelector(
		(state) => state.products.inventory,
		shallowEqual
	);
	const cart = useSelector((state) => state.products.cart);
	const cartIds = Object.keys(cart);
	const cartItems = cartIds.map((id) => (
		<CartItem key={id} id={id} quantity={cart[id].quantity} />
	));
	const total = cartIds.reduce(
		(total, id) => total + cart[id].quantity * products[id].price,
		0
	);
	const totalDiv = (
		<div className="total">
			<hr />
			<p className="has-text-right">
				<span className="subtitle">Total </span>
				<span className="title">${total}</span>
			</p>
		</div>
	);
	return (
		<div>
			{cartItems.length ? (
				[cartItems, totalDiv]
			) : (
				<p className="title my-3">No items in your cart!</p>
			)}
		</div>
	);
};

export default Cart;
