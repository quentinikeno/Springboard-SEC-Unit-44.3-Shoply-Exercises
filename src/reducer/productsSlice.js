import { createSlice } from "@reduxjs/toolkit";
import inventory from "../data.json";

const initialState = {
	inventory,
	cart: {},
};

export const productsSlice = createSlice({
	name: "products",
	initialState,
	reducers: {
		addToCart: (state, action) => {
			const { id, qty } = action.payload;
			const foundProduct = state.cart[id];
			if (foundProduct) {
				state["cart"][id]["quantity"] += qty;
			} else {
				state["cart"][id] = { quantity: qty };
			}
		},
		removeFromCart: (state, action) => {
			const { id, qty } = action.payload;
			const foundProduct = state.cart[id];
			if (foundProduct) {
				state["cart"][id]["quantity"] -= qty;
				if (state["cart"][id]["quantity"] < 1) delete state["cart"][id];
			} else {
				console.error("Item not found in cart.");
			}
		},
		clearCart: (state) => {
			state.cart = {};
		},
	},
});

export const { addToCart, removeFromCart, clearCart } = productsSlice.actions;

export default productsSlice.reducer;
