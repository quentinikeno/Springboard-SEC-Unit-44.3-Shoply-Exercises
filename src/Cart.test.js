import { render } from "@testing-library/react";
import { store } from "./reducer/store";
import { Provider } from "react-redux";
import Cart from "./Cart";

it("renders without crashing", () => {
	render(
		<Provider store={store}>
			<Cart />
		</Provider>
	);
});

it("renders matches snapshot", () => {
	const { asFragment } = render(
		<Provider store={store}>
			<Cart />
		</Provider>
	);
	expect(asFragment()).toMatchSnapshot();
});
