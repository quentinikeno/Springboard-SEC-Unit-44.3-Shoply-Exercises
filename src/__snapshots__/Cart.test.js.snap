// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <p
      class="title my-3"
    >
      No items in your cart!
    </p>
  </div>
</DocumentFragment>
`;
